packages/pymedphys_dicom/tests/test_extend_ct.py

```diff
@@ -0,0 +1,75 @@
+# Copyright (C) 2019 Cancer Care Associates
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published
+# by the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version (the "AGPL-3.0+").
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License and the additional terms for more
+# details.
+
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+# ADDITIONAL TERMS are also included as allowed by Section 7 of the GNU
+# Affero General Public License. These additional terms are Sections 1, 5,
+# 6, 7, 8, and 9 from the Apache License, Version 2.0 (the "Apache-2.0")
+# where all references to the definition "License" are instead defined to
+# mean the AGPL-3.0+.
+
+# You should have received a copy of the Apache-2.0 along with this
+# program. If not, see <http://www.apache.org/licenses/LICENSE-2.0>.
+
+
+from copy import deepcopy
+
+from pymedphys_dicom.dicom.create import dicom_dataset_from_dict
+from pymedphys_dicom.ct.extend import (
+    extend_datasets, generate_uids, convert_datasets_to_deque)
+
+
+def make_datasets(uuids, slice_locations):
+    initial_datasets = [
+        dicom_dataset_from_dict({
+            'SOPInstanceUID': uuid,
+            'InstanceNumber': str(i),
+            'SliceLocation': str(float(slice_location)),
+            'ImagePositionPatient': ['0.0', '0.0', str(float(slice_location))]
+        })
+        for i, (uuid, slice_location)
+        in enumerate(zip(uuids, slice_locations))
+    ]
+
+    return convert_datasets_to_deque(initial_datasets)
+
+
+def test_extend_datasets():
```

Comment:
@Centrus007 @msobolewski this is the primary test function making sure it is working as expected.

Reply:
Not to revive an old debate, but I'd really like to see these functions tested on actual DICOM CT files. At least once...

Reply:
So I guess a test of `ct.extend()`...

Reply:
Fair point, we have a beautiful anonimiser, which I can use to drop some files in :)

Reply:
> Not to revive an old debate, but I'd really like to see these functions tested on actual DICOM CT files. At least once...

Don't want to leave this pull request hanging for this need if that's okay?

Will assume this is a prefixed `Nit:` comment.

Reply:
So I've run into issues with the `dataset_from_dict` before. It's possible to create datasets that are valid within `pydicom` but then won't write to file.

If this is a function that is likely to result in DICOM file writes, I would really, really prefer you add file-based testing. Moreover, it really is quite a quick and easy fix.

Reply:
These tests here are primarily to test the appropriate header conversions. By design `dicom_dataset_from_dict()` in it's basic form will likely never write to file (because it needs all the extra required header info). However, these tests cover the key items of logic that these functions are undergoing.

Requiring file io for our tests when the logic has been sufficiently tested will cause our testing time to blow out. I am actually hoping in the future we can reduce the amount of file io our tests are doing.

I believe we can learn a lot from how `pydicom` itself is tested:

https://github.com/pydicom/pydicom/blob/master/pydicom/tests/test_dataset.py

It covers a lot of important cases while still having the test suite run exceptionally fast.

Reply:
Reiterating (and prettifying) some of our phone conversation here for reference :):

I'm concerned about the fact that `pydicom` permits edits to a dataset's header that would make the dataset "invalid". I.e., that the edited dataset would not successfully write to file or, worse, write but not be DICOM compliant. I'd noticed this when writing tests for DICOM anonymisation. I also noticed that this was less often the case if editing a dataset that was originally read from a file, and reasoned that there might be additional internal validity checks in `pydicom` when using datasets read from a file (e.g. mediated through the `FileDataset` class rather than `Dataset`).

For this reason, I've tended to prefer testing on datasets actually read from files (instances of `FileDataset`). Indeed, I've even tested for successful write of a DICOM file post-edits and re-read the file to check for correct edits (e.g. see [here](https://github.com/pymedphys/pymedphys/blob/4ed85f587f01e7827178b3121b1ef1d7fa27bddb/packages/pymedphys_dicom/tests/test_anonymise.py#L77) and [here](https://github.com/pymedphys/pymedphys/blob/4ed85f587f01e7827178b3121b1ef1d7fa27bddb/packages/pymedphys_dicom/tests/test_anonymise.py#L81)). I'd taken the philosophy that any function that returns an edited DICOM dataset and that will likely be written to file really ought to be tested for successful file write .

@SimonBiggs argued that file I/O is computationally expensive and in many cases probably testing overkill. I agree, especially with the former point. I then recalled that I (thought I'd) seen a `pydicom` function that explicitly checked for DICOM validity. We agreed that this would provide a suitable compromise: it avoids the risk of tests not catching invalid edits because the tests themselves work with looser constraints while also avoiding the computational code of file read/writes.

We then discovered `pydicom.config.enforce_valid_values = True`, which appears to be a configuration option that, when switched "on", adds validity checks for all DICOM dataset editing operations. If it works as we believe it does, we tentatively decided that this should simply be switched on for all testing, and possibly also for the library itself.

Still TBD:
- what exactly does `pydicom.config.enforce_valid_values = True` enforce?
- if suitable, substantially reduce DICOM file I/O in testing suite.

Reply:
Beautiful summary :)